MatplotlibDeprecationWarning examples.directory is deprecated

<!--To help us understand and resolve your issue, please fill out the form to the best of your ability.-->
<!--You can feel free to delete the sections that do not apply.-->

### Bug report

**Bug summary**

<!--A short 1-2 sentences that succinctly describes the bug-->

When running tests in `astropy` that import `matplotlib`, this warning appears, causing CI to fail when we turn warnings into errors, as shown below. I also see this warning when running the tests locally (though `pytest` complains about it, it did not result in error during test collection 🤷‍♀ ).

```
astropy/conftest.py:35: in pytest_configure
    matplotlibrc_cache.update(matplotlib.rcParams)
.../matplotlib/__init__.py:886: in __getitem__
    cbook.warn_deprecated(
.../matplotlib/cbook/deprecation.py:112: in warn_deprecated
    warnings.warn(message, category, stacklevel=2)
E   matplotlib.cbook.deprecation.MatplotlibDeprecationWarning: 
E   examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.
```

**Code for reproduction**

<!--A minimum code snippet required to reproduce the bug.
Please make sure to minimize the number of dependencies required, and provide
any necessary plotted data.
Avoid using threads, as Matplotlib is (explicitly) not thread-safe.-->

See astropy/astropy#9690

**Actual outcome**

<!--The output produced by the above code, which may be a screenshot, console output, etc.-->

Traceback as described above.

**Expected outcome**

<!--A description of the expected outcome from the code snippet-->
<!--If this used to work in an earlier version of Matplotlib, please note the version it used to work on-->

No warning emitted, as it was supposedly fixed in #12658 .

**Matplotlib version**
<!--Please specify your platform and versions of the relevant libraries you are using:-->
  * Operating system: Various
  * Matplotlib version: 2.x and 3.x
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg (on local Windows 10 machine)
  * Python version: 3.7
  * Jupyter version (if applicable): N/A
  * Other libraries: `astropy` and its test dependencies

<!--Please tell us how you installed matplotlib and python e.g., from source, pip, conda-->
<!--If you installed from conda, please specify which channel you used if not the default-->

`pip` installed into a `conda` environment (on local machine)

--- a/doc/api/next_api_changes/behavior/21042-AL.rst
+++ b/doc/api/next_api_changes/behavior/21042-AL.rst
@@ -1,0 +1,5 @@
+rcParams.copy() returns a new RcParams instance, rather than a dict
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This makes the copied instance still validate inputs, and additionally avoids
+emitting deprecation warnings when using a previously copied RcParams instance
+to update the global instance (even if some entries are deprecated).

--- a/lib/matplotlib/__init__.py
+++ b/lib/matplotlib/__init__.py
@@ -699,7 +699,10 @@ class RcParams(MutableMapping, dict):
                         if pattern_re.search(key))
 
     def copy(self):
-        return {k: dict.__getitem__(self, k) for k in self}
+        rccopy = RcParams()
+        for k in self:  # Skip deprecations and revalidation.
+            dict.__setitem__(rccopy, k, dict.__getitem__(self, k))
+        return rccopy
 
 
 def rc_params(fail_on_error=False):
@@ -877,8 +880,8 @@ dict.setdefault(rcParamsDefault, "backend", rcsetup._auto_backend_sentinel)
 rcParams = RcParams()  # The global instance.
 dict.update(rcParams, dict.items(rcParamsDefault))
 dict.update(rcParams, _rc_params_in_file(matplotlib_fname()))
+rcParamsOrig = rcParams.copy()
 with _api.suppress_matplotlib_deprecation_warning():
-    rcParamsOrig = RcParams(rcParams.copy())
     # This also checks that all rcParams are indeed listed in the template.
     # Assigning to rcsetup.defaultParams is left only for backcompat.
     defaultParams = rcsetup.defaultParams = {

--- a/lib/matplotlib/tests/test_rcparams.py
+++ b/lib/matplotlib/tests/test_rcparams.py
@@ -546,3 +546,8 @@ def test_deprecation(monkeypatch):
         mpl.rcParams["svg.hashsalt"] = "foobar"
     assert mpl.rcParams["svg.hashsalt"] == "foobar"  # Doesn't warn.
     mpl.rcParams["svg.hashsalt"] = None  # Doesn't warn.
+
+    mpl.rcParams.update(mpl.rcParams.copy())  # Doesn't warn.
+    # Note that the warning suppression actually arises from the
+    # iteration over the updater rcParams being protected by
+    # suppress_matplotlib_deprecation_warning, rather than any explicit check.

